src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs

```diff
@@ -101,66 +113,112 @@ private bool RunWorkInParallel()
+            // We grab all available workloads.
             var availableWorkloads = _workloads.Where(workload => workload != null).ToList();
+            // We take the amount of workloads to fill all the slots, or just as many workloads
+            // as there are if there are less workloads than slots.
             var amount = Math.Min(availableSlots.Count, availableWorkloads.Count);
-            var workloadsToRun = availableWorkloads.Take(amount).ToList();
+            var workloadsToAdd = availableWorkloads.Take(amount).ToList();
 
+            // We associate each workload to a slot, if we reached the max parallel
+            // level, then we will run only initalize step of the given workload.
             for (int i = 0; i < amount; i++)
             {
                 var slot = availableSlots[i];
-                slot.IsAvailable = false;
-                var workload = workloadsToRun[i];
-                workToRun.Add(new SlotWorkloadPair(slot, workload));
+                slot.HasWork = true;
+                var workload = workloadsToAdd[i];
+                var initializeOnly = occupiedSlots + i + 1 > MaxParallelLevel;
+                slot.ShouldPreStart = initializeOnly;
+
+                var manager = _createNewManager(workload.Provider);
+                var eventHandler = _getEventHandler(_eventHandler, manager);
+                slot.EventHandler = eventHandler;
+                slot.Manager = manager;
+                slot.ManagerInfo = workload.Provider;
+                slot.Work = workload.Work;
+
                 _workloads.Remove(workload);
             }
 
+            slots = _managerSlots.ToList();
             SetOccupiedSlotCount();
-
-            foreach (var pair in workToRun)
-            {
-                var manager = _createNewManager(pair.Workload.Provider);
-                var eventHandler = _getEventHandler(_eventHandler, manager);
-                pair.Slot.EventHandler = eventHandler;
-                pair.Slot.Manager = manager;
-                pair.Slot.ManagerInfo = pair.Workload.Provider;
-                pair.Slot.Work = pair.Workload.Work;
-            }
         }
 
         // Kick of the work in parallel outside of the lock so if we have more requests to run
         // that come in at the same time we only block them from reserving the same slot at the same time
         // but not from starting their assigned work at the same time.
-        foreach (var pair in workToRun)
+
+        // Kick of all pre-started hosts from the ones that had the longest time to initialize.
+        var startedWork = 0;
+        foreach (var slot in slots.Where(s => s.HasWork && !s.IsRunning && s.IsPreStarted).OrderBy(s => s.PreStartTime))
         {
-            try
+            startedWork++;
+            slot.IsRunning = true;
+            System.Diagnostics.Debug.Write($">>>>>>>>> run on pre-started host: {(DateTime.Now.TimeOfDay - slot.PreStartTime).TotalMilliseconds}ms {slot.InitTask?.Status}");
```

Comment:
Shall we keep this Debug.Write line?

Reply:
I'd keep it around for a while until we make sure nothing abnormal happens with the scheduling algorithm. We can remove it later, I can add a TODO. What do you think ?

Reply:
Are they displayed somewhere outside of the debug window? If not then it doesn't seem so useful to me. You can do EqtTrace verbose for this instead.

Reply:
Fixed, thanks !